CI: Check commit message compliance

norihiro opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->


<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
This PR adds these checks for commit messages in the logs newer than the `master` of `obsproject/obs-studio`.
- First word is starting from an upper case letter.
- The title is 50 characters at max excluding module prefix.
- Full description lines are 72 columns max.

For example, these commits are flagged as an error.
Error: commit 920a160600d6d59046082a42e90555d543cf320e: first word: FIx
Error: commit e2b7597fb0190c3b2de53201fcd246885a96e9a6: too long title: UI: Fix YT chat being shown when selecting private event
Error: commit d6f00ccfd7c4486b686bee072222c7eec298bc9f: too long title: UI: Only start YT check thread if auto-start is disabled
Error: commit f832d14220b572165bf69ad73f30a5b961d02258: too long title: UI: Disable hotkeys when a user is expected to type text

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
There are lot of violations of the commit guidelines.
This PR will enable to check the commit messages before a PR is merged.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
The script to check the guideline is tested with old commit messages and confirmed no false errors.
Unusual module name will be flagged as a warning since I don't know what is the correct module name from the history.

Also checked a wrong commit can be flagged like this. https://github.com/norihiro/obs-studio/pull/2

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented:
This is essentially a different version of #5000.  Please see the concerns listed there.

I don't see these concerns covered in the description:
* There are occasions where the commit prefix contains a slash 
* the commit message subject contains the word "Revert" before the rest of the message follows in quotes
* the commit message has no prefix
* the commit message subject has more than 50 characters after the prefix because making it shorter makes it non-descriptive and our hard limit is 72 characters total

> The title is 50 characters at max excluding module prefix.

Again, to me, the 50 character limit is not hard limit.  There are some cases where making a commit subject fit within 50 character (excluding the prefix) makes it less comprehensible than if you had used 53 characters and still came in under 72 characters total.  The 50 character limit is a suggestion or guideline that should be taken seriously, but the absolute hard limit is 72 characters total, as that is when GitHub will wrap the text.

notr1ch commented:
Agreed on the 50 characters being more of a suggestion. I frequently go over that trying to make a concise title, but definitely the 72 hard limit should apply.

norihiro commented:
The concerns mentioned by @RytoEX are addressed in the commit message and the description of this PR.
The subject is checked not to exceed 72 characters. If it exceeds 50 characters, just a warning message will be displayed but won't fail.
I wrote the script in awk. If another language is preferrable, please let me know.

RytoEX commented:
> 
> 
> The concerns mentioned by @RytoEX are addressed in the commit message and the description of this PR.
> The subject is checked not to exceed 72 characters. If it exceeds 50 characters, just a warning message will be displayed but won't fail.
> I wrote the script in awk. If another language is preferrable, please let me know.

> The subject is 72 characters at max excluding module prefix.

This is incorrect.  The hard limit on maximum length of the subject, _including the prefix_, is 72 characters.  GitHub's UI will wrap the subject if it goes beyond 72 characters.

RytoEX commented:
> Error: commit 84b4257: first word: De-escalate

"De-escalate" is technically a correct spelling of this verb.  "Deescalate" is also acceptable.  I don't know that it's easily detectable if a hyphen used in the first word is correct or not.

norihiro commented:
The title line criteria is revised.
- Hard limit to 72 characters including module prefix
- Warning at 50 characters excluding module prefix

The revised script will accept words with a single hyphen at the middle of the word such as "De-escalate". Also accept single hyphens appear multiple times such as "Click-and-drag".

Description is also updated.

WizardCM reviewed (changes requested):
So, my method for testing this PR was to grab every commit since 27.0.0 (35c07bb7e6) and run it through the script.

Of 289 recent commits

* Errors: 39
* Warnings: 26
* Info: 26

More specifically:
* 34 description lines were too long (~76/72 characters)
* 26 had titles too long (~52/50 characters)
* 26 had prefixes that were considered "not frequent"
* 4 had an "invalid" first word

Overall, I'm happy with the results and can see a benefit to having this workflow. Thank you so much for putting in the work! :)

My primary concern looking through the prefix results & the code that triggered them is that `check_module_name()` is hardcoded and a subset of the actual prefixes that may appear.

Here are the ones that it complained about:

```
UI,obs-transitions:
decklink-output-ui:
flatpak:
graphics-hook:
libobs,
libobs,deps/media-playback:
libobs-d3d11:
libobs-winrt,
libobs-winrt:
mac-capture:
mac-syphon:
mac-virtualcam:
obs-ffmpeg,
obs-frontend-api:
obs-qsv11:
text-freetype2:
```

I personally think that this check could be as simple as checking whether items in the commit prefix (comma or slash separated) match any one of the following:

* the modified file's current directory
* the modified file's root-level directory
* the modified file's name without extension
* if within a plugin subdirectory, the plugin's directory name

While I realise this is an 'Info' level warning, it can also be noisy.

Finally, for the "first word" check, I think it might be a little too aggressive:

```
first word: FIx
first word: pthread_mutex_init_recursive
first word: DrawSrgbDecompressPremultiplied
first word: add
first word: NVENC
```

Catching "FIx" and "add" is good, but function names and component names are valid first words in a commit title (in my opinion).

----

I did have issues running this script outside of the GitHub Actions workflow, and was wondering if there's something we could do about that - specifically, my `awk` is in `/usr/bin/awk` and not `/bin/awk`.

And I'd also like to see some comments in the awk script, specifically around the purposes/goals of some of the regular expressions.

RytoEX commented:
> I personally think that this check could be as simple as checking whether items in the commit prefix (comma or slash separated) match any one of the following:
> 
>     * the modified file's current directory
> 
>     * the modified file's root-level directory
> 
>     * the modified file's name without extension
> 
>     * if within a plugin subdirectory, the plugin's directory name

As far as I understand it, this script isn't checking modified files at all.  It's just checking the output of `git log` for commit messages to check the commit message prefix against a predetermined list of options.  Parsing the git log contents for modified is way more work.

In any case, what about commits that modify files in multiple directories?


> Finally, for the "first word" check, I think it might be a little too aggressive:
> 
> Catching "FIx" and "add" is good, but function names and component names are valid first words in a commit title (in my opinion).

Technically, a function name is probably against the commit guidelines unless it's treated as a verb, but there's no easy way to detect "is this being used as a verb or not".

All of these issues though continue to make me think that a lot of these either need to be non-blocking warnings, or question if the signal-to-noise ratio will be high enough to justify implementing this, per the previous PR thread.


> I did have issues running this script outside of the GitHub Actions workflow, and was wondering if there's something we could do about that - specifically, my `awk` is in `/usr/bin/awk` and not `/bin/awk`.
> 
> And I'd also like to see some comments in the awk script, specifically around the purposes/goals of some of the regular expressions.

I'll be honest, I kind of dislike that it's written in awk for a couple of reasons.  Git for Windows doesn't come with awk installed, and while it _can_ be installed on Windows, I think it's unlikely that most Windows-based devs would have it installed.  I just don't know if there's an overwhelming reason to try to use a different tool than awk, because most other tools will have similar problems, aside from my opinion that awk is perhaps a bit esoteric and complicated compared to other possible tools.

I agree that _at minimum_, this requires additional comments/documentation, because I don't expect every dev to understand how awk, or regex, works.

norihiro commented:
> I personally think that this check could be as simple as checking whether items in the commit prefix (comma or slash separated) match any one of the following:

Let me consider to retrieve to check the prefix from the changes of files.

> Technically, a function name is probably against the commit guidelines unless it's treated as a verb, but there's no easy way to detect "is this being used as a verb or not".

I thought the function names are always noun.
Maybe, detecting function names will be something like below (and possibly with numbers) but I'm afraid not all functions are covered.
- lower_case_letters_with_underscore
- CapitalizedWord

> I'll be honest, I kind of dislike that it's written in awk for a couple of reasons.

Then, how about python3? I guess more developers are familiar with python. Visual Studio includes python. Regex module (`re`) is included in it's library. There is a 3rd party's git-python library though I have never tried it yet. We may use it or just retrieving by calling `git` executable. One implementation concern is non-ascii characters, which is usually causes an error unless the environment is properly configured.

PatTheMav commented:
While I appreciate the idea in general and also the work put into this, I am not sure about added value:

Workflows do not run on all PRs automatically - this has to be triggered manually after review by a maintainer, which means that a human review has taken place yielding the same information as this check. But it is specifically PRs by non-regulars which might run afoul of this.

I share the reservations with regards to using `awk` and wonder if the same could be achieved by querying the Github API via `curl` and parsing the result via `jq` (to avoid a checkout just to check the git log), though that depends on whether we can query information for all commits associated to a PR that way.

Last, I maintain my reservation about introducing too many single-job workflows in parallel, we have too many of those running already, leading to an overcrowded "checks" area where one cannot see the forest for the trees.

norihiro commented:
I will consider your valuable comments.

norihiro commented:
I'm sorry for my late response. I've revised the script.
- The check script is rewritten in python.
- The commit history is taken using GitHub REST API.
- The module prefix is checked with the tree.

norihiro commented:
I'm afraid I accidentally clicked something and got *removed request for PatTheMav*.

norihiro commented:
Added a small modification to the script to exclude links in the commit message like below.
```text
  [1] https://www.example.com/long/long/long/long/reference-link/that-exceeds-the-72-column-rule
```

PatTheMav commented:
It would probably make more sense to convert this into a repository action and run it as part of the `check-format` workflow and (if necessary) guard it behind a condition to only run on PR pull events.

PatTheMav commented:
@norihiro Are you still willing to address the feedback? As I mentioned in my prior comment, this can absolutely live as a repository action (now that we have those) and be run in parallel with the other check actions that we run as our workflow infrastructure is now built precisely to allow for this.

I'd maybe suggest that to get this over the finish line to limit the kinds of checks we do in the first iteration:

* Limit of 72 characters per line
* Each first line matches the expression `[A-Z][\w]+: [\x20-\x7E]+`

Those two requirements should be a "low-hanging fruit" but also represent the most obvious commit message failures we have to point out in pull requests.

Once we have this in place and the checks work as expected (i.e. run only for pull requests) we can try add more checks to it.

norihiro commented:
> Each first line matches the expression [A-Z][\w]+: [\x20-\x7E]+`

I couldn't catch your point. Did you mean something like `^[A-Za-z0-9/.-]\+: [A-Z].\+` to check (1) the module name is prefixed and (2) First letter is capitalized?

> It would probably make more sense to convert this into a repository action

This is actually unclear to me. What do you refer "repository action"? Maybe, is it the [composite action](https://docs.github.com/en/actions/tutorials/create-actions/create-a-composite-action#creating-a-composite-action-within-the-same-repository) stored under `.github/actions`?

norihiro commented:
@PatTheMav, I have updated the workflow to follow other validator's implementation. Is this ok for you?

Regarding the module name checks, I think the false errors are small enough. I've updated the summary in the PR's "How Has This Been Tested" section.
There are only 48 module-name errors in the recent 5478 commits from 27.0.0 in the master branch.

For the initial deploy, I set the severity of the module-name check to warning. It won't fail the job but just reports a warning message on the log. (Probably, nobody will see if the job is not failed.)

In today's force-push, I changed these items.
- When checking the module name, relaxed the criteria.
  - Skips some of hierarchies (eg. the suffix "frontend/themes" matches "frontend/data/themes".)
  - Iterates all sub-module names in the repo.
  - Also iterates previous commit so that the removal commit also passes.
- Rebased to the master.
- Fixed most of pylint errors.

PatTheMav commented:
> > Each first line matches the expression [A-Z][\w]+: [\x20-\x7E]+`
> 
> I couldn't catch your point. Did you mean something like `^[A-Za-z0-9/.-]\+: [A-Z].\+` to check (1) the module name is prefixed and (2) First letter is capitalized?

My bad, you are of course right. I mixed up the capitalisation requirement which obviously only applies to the first word token after the colon, not the "category" which should have no such requirement because it depends on the actual category name.

> > It would probably make more sense to convert this into a repository action
> 
> This is actually unclear to me. What do you refer "repository action"? Maybe, is it the [composite action](https://docs.github.com/en/actions/tutorials/create-actions/create-a-composite-action#creating-a-composite-action-within-the-same-repository) stored under `.github/actions`?

Again you are correct, that's what I meant.

PatTheMav commented:
> I rewrote the validator so that the code is much simpler.
> 
> * Only these criterion are check.
>   
>   * Title consists of module name(s) and subject or just subject, separated by a colon followed by a single space.     Multiple module names can be separated by a comma optionally followed by a space.
>   * First word of the subject is starting from an upper case letter.
>   * The subject is not end with a period or a space.
>   * The subject and the body of the message should be within 72 columns. Links and co-author names are excluded from this check.
> * Written in bash instead of python.

Thank you for making the adjustments, will review later.
